### DISCUSSION TITLE
How to limit the maximum number of users per client

### DISCUSSION THREAD

**Discussion by jean-christophe-manciot** [keycloak/keycloak#28972] [Q&A]
I am aware that it's possible to limit the number of concurrent sessions for each user per keycloak client within a specific realm by implementing a "User session count limiter" in specific authentication flows.

Is it possible to change such configured realm limit per user?
Or is it planned?

**Comment by codespearhead** [comment]
I wonder what use case would justify this feature.

**Reply by kamaladafrica** [reply]
I have the same requirement. In my case, the use case comes from the commercial license we offer.

**Reply by janovrom** [reply]
As answered below, this can be solved by implementing custom SPI, but having it as a baseline would help with EU Cyber Resiliency Act. To cite from IEC 62443-4-2:2019:
## CR 2.7 – Concurrent session control 
### Requirement 
Components shall provide the capability to limit the number of concurrent sessions per 
interface for any given user (human, software process or device).

There are 4 levels of security levels and having this feature would make Keycloak compliant with SL-C 3 and 4.

**Comment by jean-christophe-manciot** [comment]
@codespearhead 

You're right, the question is incorrect considering my use case: the end goal is to limit the maximum number of users who can register and access a specific client (due to CPU and RAM constraints), without knowing their e-mail addresses in advance.

I suppose it is not possible to limit the number of concurrent users per client.

Despite reading the server administration guide, I could not find any way attain that goal.

Any suggestion?

**Reply by codespearhead** [reply]
No, there isn't a built-in way to do that, because that would call for more custom behavior than the default: what should happen if a user tries to log in after the maximum number of active auth sessions is reached? What should happen if a user that's already logged in starts a concurrent auth session? What should happen if one of the auth sessions expires and the user that was actively using the system is now temporarily locked out because it couldn't log into it in time?

In case you're just trying to prevent it from crashing, I'd consider spinning up a [Keycloak Cluster](https://www.keycloak.org/high-availability/introduction) with the low-end machine being just one of the nodes in it, or at least use an external Infinispan instance instead of having Keycloak use the embedded one, both of which are more future-proof solutions than creating and maintaining your custom [Keycloak Extension](https://www.keycloak.org/docs/latest/server_development/#_providers).

**Reply by jean-christophe-manciot** [reply]
Let's suppose that **there is a ```User session count limiter``` set to 1**.

> what should happen if a user tries to log in after the maximum number of active auth sessions is reached?

If there is a limit on the number of registered users/client, what you describe cannot happen.
If a user tries to register when the max is reached, it is blocked.

> What should happen if a user that's already logged in starts a concurrent auth session?

With the ```User session count limiter``` set to 1, this cannot happen.

> What should happen if one of the auth sessions expires and the user that was actively using the system is now temporarily locked out because it couldn't log into it in time?

Since the number of logged in users cannot exceed the number of registered users, this could not happen either.

> In case you're just trying to prevent it from crashing...

The CPU/RAM constraints are on the client servers (hosting the application), not on the keycloak server.

**Reply by codespearhead** [reply]
I see. That significantly simplifies the problem.

Given this, I suggest creating a custom [Authentication SPI](https://www.keycloak.org/docs/latest/server_development/#_auth_spi) to handle the authentication login and optionally extend the [Keycloak Admin REST API](https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html) so you can set the maximum number of active auth sessions dynamically.

I currently don't have the bandwidth to create a full demo, but I've crafted a prompt which, based on a quick review, should make [ChatGPT 4 generate about 90% of the necessary code](https://github.com/keycloak/keycloak/files/15112301/28972.md).

> I am implementing a session limit for a specific client in Keycloak using the embedded Infinispan cache. Provide concrete code examples for a custom Service Provider Interface (SPI) that effectively enforces this limit directly within the authenticate method during login. Additionally, show how to extend Keycloak’s REST API for dynamic configuration and demonstrate the exact UI modifications in the admin console needed to manage this setting dynamically, focusing on actual code rather than explanatory comments.

If you do that, please link the repository with your implementation here so that this discussion can be turned into an [enhancement proposal](https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md#open-a-discussion-on-a-proposed-change) like [this](https://github.com/keycloak/keycloak/discussions/categories/ideas).
